feat(avm)!: Handle Empty Calldata

MirandaWood opened the pull request:
## Empty Calldata

This is a follow-up to #17224 which handles empty calldata within the AVM. The main change is that **we no longer return 0 for empty calldata**:

```rust
   // if calldata.len() == 0 { // <-- removed
   //     0
   //  } else {
        poseidon2_hash_with_separator_slice(calldata, GENERATOR_INDEX__PUBLIC_CALLDATA)
   // }
```
When given an empty array, the AVM will just hash the separator. This is handled inside the circuit by including a special case in the `calldata` trace which indicates we have a `calldata_size` of 0:
```
//  index |  value | context_id | latch |
// -------+--------+------------+-------+
//    0   |    0   |     id     |   1   |
```
This is different from the case where we have calldata of `[ 0 ]`, which would have `index = 1` and `calldata_hash = H([sep, 0])` (with lookups confirming that the poseidon `input_len` is actually 2 and `calldata_size = 1`). I've added some comments to explain that we should never look up the `calldata` trace with an `index` of 0 unless we are looking to confirm that some `context_id` has empty calldata.

### Notes/TODOs

- The protocol will always prepend the function selector to its calldata array, so we won't ever 'see' this case in an e2e test (but we do want to handle it)

MirandaWood commented on barretenberg/cpp/src/barretenberg/vm2/constraining/relations/tx.test.cpp:
```diff
@@ -1091,4 +1097,366 @@ TEST(TxExecutionConstrainingTest, NegativePaddingChecks)
+            { C::tx_next_context_id, 5 },
+        },
+
+        // App Logic
```
Note: all rows above this are copied from a previous test so relations pass. The important part for this PR is the poseidon hashing and calldata trace checks at the end

jeanmon replied:
Could we at least have a helper to populate common trace portion?

jeanmon commented on yarn-project/stdlib/src/hash/hash.ts:
```diff
@@ -109,10 +109,6 @@ export function computeVarArgsHash(args: Fr[]): Promise<Fr> {
  * @returns Hash of the calldata.
  */
 export function computeCalldataHash(calldata: Fr[]): Promise<Fr> {
-  if (calldata.length === 0) {
```
Did you check that both noir and TS routines support an empty array as input of poseidon?

MirandaWood replied:
I haven't actually - but in all these cases we would just hash the separator, so the poseidon input is an array of one field. However I will run some tests passing in empty arrays to these methods to make sure!

jeanmon reviewed (approved):
LGTM and great testing.
Would you mind adding a test with empty calldata in data_copy.test.cpp.
IIRC the convention is that a calladatacopy which goes beyond the end of the calldata would return FF(0) elements.

dbanks12 commented on barretenberg/cpp/pil/vm2/calldata.pil:
```diff
@@ -12,6 +12,15 @@ include "calldata_hashing.pil";
 // The size and hash of the calldata is constrained by calldata_hashing.pil. These values should be
 // looked up via calldata_hashing.pil's latch.
 //
+// For empty calldata, we have a special case where no value exists, but should have a row in this
+// trace to indicate that a certain context_id has been processed:
+//  index |  value | context_id | latch |
+// -------+--------+------------+-------+
+//    0   |    0   |     id     |   1   |
+// This is the only case in which index = 0 and sel = 1. Note that no lookups should access this trace
+// with index = 0 unless to confirm that the calldata is empty (data_copy and calldata_hashing lookup
+// all existing values with a +1).
+//
 // e.g. calldata: [0x111, 0x222, 0x333, 0x444]
```
It's cool that we didn't need to change any constraints to support this, only sim & tracegen!

One thing that feels concerning to me is just that... the circuit doesn't care that we have this row with `index = 0, sel = 1` despite the fact that we're supposed to start at `index = 1`.... Could someone do something weird by starting at some other random index instead of 0 or 1?

MirandaWood replied:
I thought a little bit about this when writing out the relationship between `calldata.pil` and `calldata_hashing.pil` - there are a few aspects the `calldata` trace doesn't care about which are constrained by the hashing trace, one of which is the start index.

The hashing trace always starts at 0 and so looks up index 1 as the first value. Since the incrementing index is constrained, a calldata instance starting somewhere other than 1 would never be constrained by the hashing trace. However, it's still possible for the values to be filled, which is weird. I don't think it's dangerous but I'm not 100%, so I'd be happy to add some starting columns + relations!

MirandaWood replied:
Discussed in call - for now adding more clarifying comments that we are constrained, with examples of how to constrain if we wish to change in future.

MirandaWood committed:
feat(avm)!: Handle Empty Calldata

## Empty Calldata

This is a follow-up to #17224 which handles empty calldata within the AVM. The main change is that **we no longer return 0 for empty calldata**:

```rust
   // if calldata.len() == 0 { // <-- removed
   //     0
   //  } else {
        poseidon2_hash_with_separator_slice(calldata, GENERATOR_INDEX__PUBLIC_CALLDATA)
   // }
```
When given an empty array, the AVM will just hash the separator. This is handled inside the circuit by including a special case in the `calldata` trace which indicates we have a `calldata_size` of 0:
```
//  index |  value | context_id | latch |
// -------+--------+------------+-------+
//    0   |    0   |     id     |   1   |
```
This is different from the case where we have calldata of `[ 0 ]`, which would have `index = 1` and `calldata_hash = H([sep, 0])` (with lookups confirming that the poseidon `input_len` is actually 2 and `calldata_size = 1`). I've added some comments to explain that we should never look up the `calldata` trace with an `index` of 0 unless we are looking to confirm that some `context_id` has empty calldata.

### Notes/TODOs

- The protocol will always prepend the function selector to its calldata array, so we won't ever 'see' this case in an e2e test (but we do want to handle it)
